pkg/actuators/machine/instances.go

```diff
@@ -505,3 +488,107 @@ func getInstanceMarketOptionsRequest(providerConfig *machinev1.AWSMachineProvide
+	}
+	if machineProviderConfig.Placement.PartitionNumber != 0 {
+		placement.SetPartitionNumber(int64(machineProviderConfig.Placement.PartitionNumber))
+	}
+
+	if *placement == (ec2.Placement{}) {
+		// If the placement is empty, we should just return a nil so as not to pollute the RunInstancesInput
+		return nil, nil
+	}
+
+	return placement, nil
+}
+
+func checkOrCreatePlacementGroup(client awsclient.Client, placement machinev1.Placement, clusterID string) error {
+	if placement.GroupName == "" {
+		// Nothing to do if the group name is empty
+		return nil
+	}
+
+	placementGroups, err := client.DescribePlacementGroups(&ec2.DescribePlacementGroupsInput{
+		GroupNames: []*string{aws.String(placement.GroupName)},
+	})
+	if err != nil && !isAWS4xxError(err) {
+		// Ignore a 400 error as AWS will report an unknown placement group as a 400.
+		return fmt.Errorf("could not describe placement groups: %v", err)
+	}
+
+	if len(placementGroups.PlacementGroups) == 1 {
+		// This is the normal path, the named placement group exists
+		return nil
+	}
+	if len(placementGroups.PlacementGroups) > 1 {
+		return fmt.Errorf("expected 1 placement group for name %q, got %d", placement.GroupName, len(placementGroups.PlacementGroups))
+	}
+
+	// No placement group by that name existed, so we create one
+	createPlacementGroupInput := &ec2.CreatePlacementGroupInput{
+		GroupName: aws.String(placement.GroupName),
+		TagSpecifications: []*ec2.TagSpecification{
+			{
+				ResourceType: aws.String(ec2.ResourceTypePlacementGroup),
+				Tags: []*ec2.Tag{
+					{Key: aws.String("kubernetes.io/cluster/" + clusterID), Value: aws.String("owned")},
+					{Key: aws.String("Name"), Value: aws.String(placement.GroupName)},
+				},
+			},
+		},
+	}
+	switch placement.GroupType {
+	case machinev1.AWSSpreadPlacementGroupType, "":
+		createPlacementGroupInput.SetStrategy(ec2.PlacementStrategySpread)
+	case machinev1.AWSClusterPlacementGroupType:
+		createPlacementGroupInput.SetStrategy(ec2.PlacementStrategyCluster)
+	case machinev1.AWSPartitionPlacementGroupType:
+		createPlacementGroupInput.SetStrategy(ec2.PlacementStrategyPartition)
+	default:
+		return fmt.Errorf("unknown placement strategy %q: valid values are %s, %s, %s and omitted", placement.GroupType, machinev1.AWSSpreadPlacementGroupType, machinev1.AWSClusterPlacementGroupType, machinev1.AWSPartitionPlacementGroupType)
+	}
+	if placement.PartitionCount != 0 {
+		createPlacementGroupInput.SetPartitionCount(int64(placement.PartitionCount))
+	}
+
+	if _, err := client.CreatePlacementGroup(createPlacementGroupInput); err != nil {
+		return fmt.Errorf("unable to create placement group: %v", err)
+	}
+
+	return nil
+}
+
+// isAWS4xxError will determine if the passed error is an AWS error with a 4xx status code.
+func isAWS4xxError(err error) bool {
+	if _, ok := err.(awserr.Error); ok {
+		if reqErr, ok := err.(awserr.RequestFailure); ok {
+			if strings.HasPrefix(strconv.Itoa(reqErr.StatusCode()), "4") {
```

Comment:
do we need to convert the int to ascii here, couldn't we just check `StatusCode() >= 400 && StatusCode() < 500`?

Reply:
This was just copied from https://github.com/openshift/machine-api-provider-aws/pull/16/files#diff-c5079a7c5d2916c3661d0c44d1267050b3496c40b4530363d3825d66af581b6fR376, but you're right, that would make more sense, I'll update it

Reply:
no worries, i just thought converting to ascii seemed like a needless use of cycles